### DISCUSSION TITLE
How to access DB from background isolate in flutter?

### DISCUSSION THREAD

**Discussion by azeem-io** [simolus3/drift#3249] [Q&A] [answered]
Hi, apologies if this has been answered before, I'm just getting confused seeing all the previous discussions, many of which refer to previous versions.

I am using android_alarm_manager_plus to create a background isolate in flutter. What is the recommended way to access the database from that isolate? And would there be any issue if I am trying to write to the database from multiple isolates?

**Comment by simolus3** [comment, marked as answer]
So generally, the best approach is to use `drift_flutter` for that, similar to what is being done here: https://github.com/simolus3/drift/discussions/3229

So as long as the two isolates share a `IsolateNameServer` instance (I think they do since they're running in the same process), you can use the `driftDatabase` function from `package:drift_flutter` with the `shareAcrossIsolates: true` option) like this:

```dart
@DriftDatabase(...)
class AppDatabase extends _$AppDatabase {
  AppDatabase() : super(driftDatabase(name: 'app.db', native: DriftNativeOptions(shareAcrossIsolates: true)));
```

Then, regardless of the isolate you open your database in, they'll all talk to the same database instance. A downside is that this is currently broken for isolates in different Flutter engines (see the linked discussion for the workaround until the upcoming drift release).

> And would there be any issue if I am trying to write to the database from multiple isolates?

In this setup, there wouldn't be any issues with that since the `shareAcrossIsolates` coordinates writes coming in from multiple client isolates (like the UI isolate or the alarm manager's isolate).

Instead of using `shareAcrossIsolates`, you can also open your database in each isolate independently. This has the downsides that stream queries aren't synchronized and that you may have to retry writes to avoid running into locks set by the other isolate (not too hard, this essentially involves setting the journey mode to WAL and enabling the `busy_timeout` pragma in sqlite3). Depending on your setup, these downsides might not be a problem at all (in which case using independent database connections to the same file might slightly speed things up because drift doesn't have to pessimistically coordinate writes from different isolates).
